### PR TITLE
Make #[derive(Copy, Eq, Ord)] imply #[derive(Clone, PartialEq, PartialOrd)]

### PR DESCRIPTION
**Pull request by erickt**
This reduces the amount of boilerplate users need to do in order to derive some common traits.

This is related to #23860.

cc @nikomatsakis, @aturon

### CODE REVIEW

File: src/libsyntax/ext/deriving/cmp/ord.rs [hunk 1]
```diff
@@ -20,114 +18,75 @@ use ext::deriving::generic::ty::*;
...
     };
-    trait_def.expand(cx, mitem, item, push)
-}
 
-#[derive(Copy)]
-pub enum OrderingOp {
-    PartialCmpOp, LtOp, LeOp, GtOp, GeOp,
+    trait_def.expand(cx, mitem, item, push);
+
+    partial_ord::expand_deriving_partial_ord(cx, span, mitem, item, push)
```
Comment (sfackler) [inline src/libsyntax/ext/deriving/cmp/ord.rs]: The derived `partial_cmp` can be optimized in this case to just `Some(self.cmp(other))`.

**Comment by aturon** [comment]
:dart:

File: src/libsyntax/ext/deriving/cmp/ord.rs (hunk 1 above)
Comment (erickt) [inline src/libsyntax/ext/deriving/cmp/ord.rs]: Good point. I've got a build running that applies this PartialOrd optimization to non-generic types. I'll push it up when it finishes running.

**Comment by alexcrichton** [comment]
Out of curiosity, is it still possible to have `#[derive(Ord)]` but a manual implementation of `PartialOrd`?

**Comment by sfackler** [comment]
@alexcrichton I think the more common thing would be the opposite since `PartialOrd` has a trivial impl if you've already implemented `Ord::cmp`.

**Comment by alexcrichton** [comment]
Hm true. In general I've been very worried in the past about this form of `deriving` as it seems like a footgun for various forms of configuration of `deriving`. For example it looks like instances of `derive(Eq)` with a manual `PartialEq` impl were removed in favor of a manual `Eq` impl. I worry about the "half-`derive`" case, but it may not be too large of a worry in practice. For example adding an impl of `Eq` is just a marker trait so it's not so bad, and I'm not sure how much this applies to `Ord` and `PartialOrd`...

**Comment by erickt** [comment]
@alexcrichton: with this PR we could no longer provide a manual implementation of `PartialOrd` with `#[derive(Ord)]`. Is there ever a case where we'd want a manual implementation if we already have version of `Ord`? That also seems like it could be a source of bugs if the two were inconsistent.

**Comment by nikomatsakis** [comment]
My main concern is whether this PR requires an RFC. It seems to be at least mildly controversial. Obviously this would have to land by 1.0 though!

(I personally acknowledge @alexcrichton but think that this PR is a net improvement on the current situation.)

**Comment by erickt** [comment]
@sfackler: I just pushed up a patch that simplifies generating `PartialOrd` for non-generic `Ord` types. I'm not sure if my metrics are right, but this PR appears to be shaving off about 100KB from `libstd.dylib` when compared with the latest nightly `libstd.dylib`, but I can't say if that's all due to this PR or some other optimizations going on. I could do a fresh build to compare the too if we want a real number.

**Comment by nikomatsakis** [comment]
r+ from me (modulo RFC) but I'm not really *that* familiar with this code. Certainly I'd like to hear @sfackler or @huonw weigh in.

**Comment by erickt** [comment]
I rebased on top of HEAD to avoid some merge conflicts and squashed some of the cleanup commits together.

**Comment by huonw** [comment]
Copying [this comment](https://github.com/erickt/rust/commit/eed443b25c8b52c2ab45c777b50c180d7db6b414#commitcomment-10559139) to ensure it doesn't get lost: I believe it one should be able to make the simple-impls work via the `self_args` fields of Substructure.

Also, another minor point: what about supporting `#[derive(Copy, Clone)] struct Foo;` (i.e. explicitly write the `Clone`?)?

Also, I wonder about the case of optimising `PartialOrd` (e.g. maybe `<` can be computed faster than a full comparison). There's a big cliff between:

```rust
#[derive(Ord)]
struct Foo { ... }
```

and

```rust
struct Foo { ... }

impl Ord for Foo {
    fn cmp(&self, other: &Foo) -> Ordering {
        // ... lots of stuff ...
    }
}
impl PartialOrd for Foo {
     fn partial_cmp(&self, other: &Foo) -> Option<Ordering> { Some(self.cmp(other)) }
     fn lt(&self, other: &Foo) -> bool { /* supafast */ }
}
```

That said, it seems like this PR is the right default, and we can add an opt-out mechanism in future.

**Comment by nikomatsakis** [comment]
> Also, another minor point: what about supporting #[derive(Copy, Clone)] struct Foo; (i.e. explicitly write the Clone?)?

I thought the PR did permit this, albeit with a FIXME attached? (I tend to think we should just allow it and, at most, lint for it)

**Comment by erickt** [comment]
@nikomatsakis: Yes, this PR explicitly handles `#[derive(Copy, Clone)]`. A lint sounds fine to me.

**Comment by huonw** [comment]
Oh, I wasn't clear: my reading of the PR was that the support for that was only for staging reasons. In any case, a lint (or nothing) sounds good to me.

**Comment by dcrewi** [comment]
This should probably be closed now, as the RFC has been resolved?

**Comment by alexcrichton** [comment]
Yes thanks for the reminder @dcrewi! Closing for the same reasons [as listed in the RFC](https://github.com/rust-lang/rfcs/pull/1028#issuecomment-91566317).

**Pull request closed without merging**